### ISSUE TITLE
Unecessary Warning message

### ISSUE THREAD

**Issue by andrewasd** [rust-lang/rust#85913]
I tried this code:

```rs
#![allow(dead_code)]

fn fun() -> i32 {
    function() && return 1;

    panic!("stop");
}

fn function() -> bool {
    true
}

fn main() {}

```

where `function` is any function that returns a `boolean` , and it gives  the warning
*unused logical operator* , but actually works as expected, so i think this warning should not come up.

**Comment by asquared31415** [comment]
The expression `a && return val` produces a value, though you will never observe it to be true, because if it is, it will return early.  The logical operators are all set up so that it warns if they are not used because in most cases the programmer means to use the value for something.  You are not actually using the returned value, since you never read it or assign it to any variable.  An explicit discard `let _ = a && return val;` stops the warning because you indicate that you don't want to use the value and this is not a mistake.

**Comment by andrewasd** [comment]
@asquared31415  thank you

**Comment by HKalbasi** [comment]
Is my issue #88017 duplicate of this issue?
